adagrams/game.py

```diff
@@ -1,11 +1,104 @@
 def draw_letters():
-    pass
+    import random
+    LETTER_POOL = {
+    'A': 9, 
+    'B': 2, 
+    'C': 2, 
+    'D': 4, 
+    'E': 12, 
+    'F': 2, 
+    'G': 3, 
+    'H': 2, 
+    'I': 9, 
+    'J': 1, 
+    'K': 1, 
+    'L': 4, 
+    'M': 2, 
+    'N': 6, 
+    'O': 8, 
+    'P': 2, 
+    'Q': 1, 
+    'R': 6, 
+    'S': 4, 
+    'T': 6, 
+    'U': 4, 
+    'V': 2, 
+    'W': 2, 
+    'X': 1, 
+    'Y': 2, 
+    'Z': 1
+    }
+    draw_pool = []
+    for letter, frequency in LETTER_POOL.items():
+        for i in range(frequency):
+          draw_pool.append(letter)
+    hand = []
+    for i in range(10): # wish I had known about .sample before submitting!
+        piece_index = random.randint(0,len(draw_pool)-1)
+        piece = draw_pool[piece_index]
+        hand.append(piece)
+        draw_pool.remove(piece)
+    return hand
 
 def uses_available_letters(word, letter_bank):
-    pass
+    new_word = word.upper()
+    copy_bank = letter_bank.copy()
+    for letter in new_word:
+      if letter in copy_bank:
+        copy_bank.remove(letter)
```

Comment:
Notice here that you need to loop through every letter in new_word on line 46. Then you're using the in operator to check if the letter is in the copy_bank on line 47. Recall that using the `in` operator has O(n) time complexity, as does your for loop, so we've got O(n^2).

Rather than using a list of the letters in the hand, could we build a helper data structure (like a dictionary) that could let us look up the number of tiles remaining of each type?

If you have a frequency dict and then check if a key is in the dict, then you can leverage the fact that dicts constant time lookup for keys (in contrast to the list's linear time look up).